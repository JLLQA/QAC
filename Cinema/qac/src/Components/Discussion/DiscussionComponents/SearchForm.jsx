import { Form } from "reactstrap";

const SearchForm = (props) => {
    const { handleSubmitSearch, query } = props;
    return (
        <Form>
            <input id="search" name="search " type="text" placeholder="Search by topic title" value={query} onChange={handleSubmitSearch} />
        </Form>
    )
}

export default SearchForm;