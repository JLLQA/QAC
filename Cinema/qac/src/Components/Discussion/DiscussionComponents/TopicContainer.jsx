
import React from 'react';
import { Link } from 'react-router-dom';
import { Card, CardText, CardBody, CardTitle, CardSubtitle, Container, Row, Col } from 'reactstrap';

const TopicContainer = (props) => {
    const { data, query, refresh } = props

    const Filter = require("bad-words");
    const filter = new Filter();

    if (query.length > 0) {
        return (
            <Container align="center">
                {data
                    .filter(post => post.title.toLowerCase().includes(query.toLowerCase()))
                    .map((post) => (
                        <Row key={post.id}>
                            <Col>
                                <Card >
                                    <CardBody>
                                        <CardTitle tag="h5">{filter.clean(post.title)}</CardTitle>
                                        <CardSubtitle tag="h6">{filter.clean(post.username)}</CardSubtitle>
                                        <CardText>{filter.clean(post.body)}</CardText>
                                        <Link to={{ pathname: `/discussion/${post.title}` }}>
                                            <h3>More</h3>
                                        </Link>
                                        <br />
                                    </CardBody>
                                </Card>
                                <br />
                            </Col>
                        </Row>
                    ))}
            </Container>
        )
    } else {
        return (
            // axios get request for all topic objects - map instead of dummy data
            <Container align="center">
                {data.map((post) => (
                    <Row key={post.id}>
                        <Col>
                            <Card>
                                <CardBody>
                                    <CardTitle tag="h5">{filter.clean(post.title)}</CardTitle>
                                    <CardSubtitle tag="h6">{filter.clean(post.username)}</CardSubtitle>
                                    <CardText>{filter.clean(post.body)}</CardText>
                                    <Link to={{ pathname: `/discussion/${post.title}` }}>
                                        <h3>More</h3>
                                    </Link>
                                </CardBody>
                            </Card>
                            <br />
                        </Col>
                    </Row>
                ))}
            </Container>
        )
    }

}

export default TopicContainer;