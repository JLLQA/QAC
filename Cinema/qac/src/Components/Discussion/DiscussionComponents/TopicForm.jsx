import { Col, Form } from "reactstrap";

const TopicForm = (props) => {
    const { title, username, body, handleSubmitTopic, handleUsername, handleTitle, handleBody } = props;
    return (
        <Col align="center">
            <Form>
                <input type="text" name="username" id="username" placeholder="Username" value={username} onChange={handleUsername} />
                <br />
                <input type="text" name="topic_title" id="topic_title" placeholder="Topic Title" value={title} onChange={handleTitle} />
                <br />
                <input type="text" name="topic_body" id="topic_body" placeholder="Topic Body" value={body} onChange={handleBody} />
                <br />
                <br />
                <button id="submitButton" onClick={handleSubmitTopic} type="submit">Create Topic</button>
                <br />
                <br />
            </Form>
        </Col>
    )
}

export default TopicForm;