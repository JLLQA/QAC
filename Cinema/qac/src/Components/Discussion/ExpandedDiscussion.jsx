import { Col, Container, Row } from "reactstrap";

const ExpandedDiscussion = (props) => {
    const {comments,title,body,username} = props
    return(
        <Container>
            <Row>

            </Row>
            <Row>
                {comments.map((comment) => {
                    <Col>
                        <h1>Username: {comment.username}</h1>
                        <h2>Comment: {comment.body}</h2>
                        
                    </Col>
                })}
            </Row>
        </Container>
    )
}

export default ExpandedDiscussion;