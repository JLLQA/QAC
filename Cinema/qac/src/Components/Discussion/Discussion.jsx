import { useEffect, useState } from "react";
import { Col, Container, Form, Nav, Row } from "reactstrap"
import axios from "axios";
import Navbar from "../Navbar/Navbar"
import SearchForm from "./DiscussionComponents/SearchForm";
import TopicContainer from "./DiscussionComponents/TopicContainer";
import TopicForm from "./DiscussionComponents/TopicForm";

const Discussion = () => {
    const [query, setQuery] = useState("");
    const [title, setTitle] = useState("");
    const [username, setUsername] = useState("");
    const [body, setBody] = useState("");

    const handleSubmitSearch = (event) => {
        setQuery(event.target.value);
    }
    const handleTitle = (event) => {
        setTitle(event.target.value);
    }
    const handleUsername = (event) => {
        setUsername(event.target.value);
    }
    const handleBody = (event) => {
        setBody(event.target.value);
    }

    const handleSubmitTopic = (event) => {
        event.preventDefault();
        console.log(title);
        console.log(username);
        console.log(body);
        //axios request to create a new discussion object using states
    }

    const [data, setData] = useState([]);
    const [isLoaded, setIsLoaded] = useState(false);

    useEffect(() => {
        axios.get(`http://www.omdbapi.com/?i=tt3896198&apikey=14ff12ef`)
            .then((res) => {
                setData(res.data);
                setIsLoaded(true);
            }).catch((err) => {
                console.log(err.message);
                setIsLoaded(true);
            });
    }, []);

    if (isLoaded) {
        return (
            <div>
                <Navbar />
                <Container align="center">
                    <h1 style={{ color: "White" }}>Discussion</h1>
                    <Row>
                        <Col align="center">
                            <SearchForm
                                handleSubmitSearch={handleSubmitSearch}
                                query={query}
                            />
                        </Col>
                    </Row>
                    <Row>
                        <TopicForm
                            title={title}
                            username={username}
                            body={body}
                            handleSubmitTopic={handleSubmitTopic}
                            handleUsername={handleUsername}
                            handleTitle={handleTitle}
                            handleBody={handleBody}

                        />
                    </Row>
                </Container>
                <TopicContainer
                    data = {data}
                />
            </div>
        )
    } else {
        return(
            <>
            <Navbar/>
            <h1>SOMETHING WRONG</h1>
            </>
        )

    }
}



export default Discussion;