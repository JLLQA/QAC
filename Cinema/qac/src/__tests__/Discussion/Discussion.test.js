import '@testing-library/jest-dom';
import axios from "axios";
import { fireEvent, render, screen } from '@testing-library/react';
import { BrowserRouter as Router } from "react-router-dom"
import Discussion from '../../Components/Discussion/Discussion';
import SearchForm from '../../Components/Discussion/DiscussionComponents/SearchForm';
import TopicForm from '../../Components/Discussion/DiscussionComponents/TopicForm';
import { shallow } from 'enzyme';
import { useState } from 'react';
import renderer from "react-test-renderer"

import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import TopicContainer from '../../Components/Discussion/DiscussionComponents/TopicContainer';
Enzyme.configure({ adapter: new Adapter() });


describe("testing discussion page functionality", () => {


    test('Test click event', () => {
        const mockCallBack = jest.fn();

        const button = shallow((<button id="submitButton" onClick={mockCallBack} type="submit">Create Topic</button>));
        button.find('button').simulate('click');
        expect(mockCallBack.mock.calls.length).toEqual(1);
    });

    test("testing links", async () => {
        const mockdata = [{
            id: "1",
            title: "title",
            username: "username",
            body: "body"
        }]
        render(<Router>
            <TopicContainer
                query=""
                data={mockdata}
            />
        </Router>);
        expect(screen.getByRole('link', { name: /More/i })).toHaveAttribute('href', '/topics/title');
    });

    //backend server 5000 has to be running
    test("Discussion text is present", async () => {
        render(
            <Router>
                <Discussion />
            </Router>
        );
        await axios.get(`http://localhost:5000/topics`);

        const discussionText = screen.getByText("DISCUSSION BOARD");
        expect(discussionText).toBeInTheDocument();
    });


    test("Loading text is present when data has not loaded", () => {
        render(
            <Router>
                <Discussion />
            </Router>
        );
        const loadingText = screen.getByText("Loading...");
        expect(loadingText).toBeInTheDocument();
    });


    test("Text component loaded correctly", () => {
        render(
            <Router>
                <Discussion />
            </Router>
        );
        const tree = renderer
            .create(<h1>DISCUSSION BOARD</h1>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });


    test("Testing if search bar passes values correctly", () => {
        const page =
            render(

                <Router>
                    <SearchForm />
                </Router>
            );
        const input = page.container.querySelector("#search");
        fireEvent.change(input, { target: { value: "testingthisreturnsnothing" } });
        expect(input.value).toBe("testingthisreturnsnothing");
    })

    test("Testing if topic inputs changes value correctly", () => {
        const page = render(
            <Router>
                <TopicForm />
            </Router>
        );
        const input1 = page.container.querySelector("#username");
        const input2 = page.container.querySelector("#topic_title");
        const input3 = page.container.querySelector("#topic_body");
        fireEvent.change(input1, { target: { value: "testingthisreturnsnothing" } });
        fireEvent.change(input2, { target: { value: "testingthisreturnsnothing" } });
        fireEvent.change(input3, { target: { value: "testingthisreturnsnothing" } });
        expect(input1.value).toBe("testingthisreturnsnothing");
        expect(input2.value).toBe("testingthisreturnsnothing");
        expect(input3.value).toBe("testingthisreturnsnothing");
    })





})