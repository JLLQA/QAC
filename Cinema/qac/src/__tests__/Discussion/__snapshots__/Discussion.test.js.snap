// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing discussion page functionality Text component loaded correctly 1`] = `
<h1>
  DISCUSSION BOARD
</h1>
`;
